openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiMapSchemaSpec.groovy

```diff
@@ -0,0 +1,121 @@
+
+    void "test build OpenAPI doc for map properties"() {
+
+        when:
+        buildBeanDefinition('test.MyBean', '''
+package test;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import io.micronaut.core.annotation.Introspected;
+import io.micronaut.http.HttpResponse;
+import io.micronaut.http.MediaType;
+import io.micronaut.http.annotation.Body;
+import io.micronaut.http.annotation.Controller;
+import io.micronaut.http.annotation.Post;
+import io.reactivex.Single;
+import io.swagger.v3.oas.annotations.Operation;
+import io.swagger.v3.oas.annotations.media.Content;
+import io.swagger.v3.oas.annotations.media.Schema;
+import io.swagger.v3.oas.annotations.parameters.RequestBody;
+import io.swagger.v3.oas.annotations.responses.ApiResponse;
+import io.swagger.v3.oas.annotations.tags.Tag;
+
+import io.micronaut.core.annotation.Nullable;
+import jakarta.validation.Valid;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Min;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Represents a pet.
+ */
+@Introspected
+abstract class Pet {
+
+    private Map<String, String> map;
+
+    private Map<String, Pet> complexMap;
+
+
+    public Pet(Map<String, String> map, Map<String, Pet> complexMap) {
+        this.map = map;
+        this.complexMap = complexMap;
+    }
+
+    public Pet() {
+    }
+
+    public Map<String, String> getMap() {
+        return map;
+    }
+
+    public Map<String, Pet> getComplexMap() {
+        return complexMap;
+    }
+
+    public void setMap(Map<String, String> map) {
+        this.map = map;
+    }
+
+    public void setComplexMap(Map<String, Pet> complexMap) {
+        this.complexMap = complexMap;
+    }
+}
+
+@Controller("/pets")
+class PetController {
+
+    @Operation(summary = "Save a pet", description = "The saved pet information is returned")
+    @Tag(name = "save-pet")
+    @Post
+    HttpResponse<Pet> save(@Valid @NotNull @Body Pet pet) {
+        return HttpResponse.ok(pet);
+    }
+}
+
+''')
+        then: "the state is correct"
+        Utils.testReference != null
+
+        when: "The OpenAPI is retrieved"
+        OpenAPI openAPI = Utils.testReference
+        Schema petSchema = openAPI.components.schemas['Pet']
+
+        then: "the components are valid"
+        petSchema != null
+
+        petSchema instanceof Schema
+
+        petSchema.type == 'object'
+        petSchema.properties.size() == 2
+        petSchema.properties['map'] instanceof MapSchema
+        petSchema.properties['complexMap'] instanceof MapSchema
+
+        when:
+        MapSchema map = (MapSchema) petSchema.properties['map']
+        MapSchema complexMap = (MapSchema) petSchema.properties['complexMap']
+
+        then:
+        map.type == 'object'
```

Comment:
For some reason this test didn't catch the issue, but when I inspected the created `OpenAPI` object in `OpenApiApplicationVisitor` ([here](https://github.com/micronaut-projects/micronaut-openapi/blob/25f05ad73d2db77722b2a6917bf576a63bcbef0a/openapi/src/main/java/io/micronaut/openapi/visitor/OpenApiApplicationVisitor.java#L501)) the map property had the incorrect type.

I am not sure how this schema is different from the one that is there. @altro3 do you know?

Reply:
I can only watch it tomorrow. But, as the ticket author wrote, the problem has already been resolved in version 5.1.0.

I can add that the generated OpenApi (to which you put down properties throughout the entire logic) and the one that will be in the final are different objects, because this object is serialized and then deserialized using Jackson. Those. some of the properties may differ from the original object after deserialization according to the swagger rules.